### PR TITLE
feat(UserStatus): Add user status in real time - 118

### LINKED ISSUE #118: Profile: add user status
### Already available User Status in warp-dart to add on Uplink

image

### PR DESCRIPTION
**Pull request by lgmarchi**
### 1. Add user status in real time with streams 
a. Possible to see user status working on Add Friend Page and Chat Room Page 
b. Contacts lists will added in other PR, because the logic need to be different 
c. For now is possible to see just one user with user status working
d. Status is online when other user is with app opened, status is offline when other user close the app
https://user-images.githubusercontent.com/63157656/197244512-a6c31d92-e601-4699-b793-42c8fe730cf7.mov

### 2. Stream watch user status
a. This is a stream on Uplink side, so it was necessary to create a Stream producer with while loop
image
b. To be possible to stop stream, it was created as "side channel" and call a function to stop while loop
image
c. New data will be just yielded when it is different from past value
image

### 3. Stream on Bloc 
a. To listen all new event on stream, it was used emit.forEach, that will run when new data arrive and emit new state 
image
b. On this event, it is setted as restarbable event, because it will stop past event when a new one is called, this function come from [bloc_concurrency package](https://pub.dev/packages/bloc_concurrency)
image
c. To start stream, it is necessary call event, to stop it call function to close stream
image
image
image

### 4. Stream on UI 
a. Add Friend Page
b. Chat Room Page, here if user comes from Add Friend Page, it will just close stream on Add Friend Page, if comes from User Profile Page, the stream will be closed when the page is closed

### 5. Profile Index Page 
a. There was an old error, that when user delete avatar or banner picture, it didn't go away on UI, problem fixed 
b. Other problem was when we don't have avatar and banner picture, the profile picture widget need to have a black border to separate it from banner picture, and depending on the situation, this border disappear, problem fixed

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [x] 🧹 Code refactor

### CODE REVIEW

**Comment by phillsatellite** [comment]
Hi Lucas! Function of this PR works great! I found this little bug when testing and wasnt sure if we wanted to fix it on here or in a separate PR but I noticed when user adds and friend and goes the message them, the first time the screen will be grey if they go back and enter messages again will appear fine 

https://user-images.githubusercontent.com/93608357/197568719-93f9f052-1f10-4cda-a6df-a72e74d1c118.mp4

**Review by phillsatellite — APPROVED** [review]

File: lib/contacts/presentation/view/add_friend_page/add_friend_page.dart [hunk 1]
```diff
@@ -42,10 +42,13 @@ class _AddFriendPageState extends State<AddFriendPage> {
...
         }
         if (_textController.text != oldIdToSearch) {
-          _friendController.add(ResetFriendDataStarted());
+          _friendController
+            ..add(ResetFriendDataStarted())
+            ..closeWatchUserStatusStream();
         }
       });
     });
+    super.initState();
```
Comment (hyj1204) [inline lib/contacts/presentation/view/add_friend_page/add_friend_page.dart]: move super.initState() to the top

**Review by hyj1204 — APPROVED** [review]
Good Job. The user status in AddFriendPage and ChatRoomPage work well. 
AddFriendPage 

https://user-images.githubusercontent.com/14248245/198399660-53693a17-21d1-4f61-bf6f-118fbad5c6f5.mov


ChatRoomPage 

https://user-images.githubusercontent.com/14248245/198399534-c8f5b33c-472f-4ea3-9983-1f683a782cf2.mov

File: lib/contacts/presentation/view/add_friend_page/add_friend_page.dart (hunk 1 above)
Comment (lgmarchi) [reply lib/contacts/presentation/view/add_friend_page/add_friend_page.dart]: Done.

**Pull request merged**